<a href="https://www.kaggle.com/code/ayusha232/problem-5-sml?scriptVersionId=106926371" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
from sklearn.datasets import load_diabetes
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

In [2]:
 diabetes_X, diabetes_y = load_diabetes(return_X_y=True,as_frame=True )

In [3]:
 diabetes_X = diabetes_X[['age','sex','bmi','bp']]

In [4]:
 diabetes_X

age       sex       bmi        bp
0    0.038076  0.050680  0.061696  0.021872
1   -0.001882 -0.044642 -0.051474 -0.026328
2    0.085299  0.050680  0.044451 -0.005671
3   -0.089063 -0.044642 -0.011595 -0.036656
4    0.005383 -0.044642 -0.036385  0.021872
..        ...       ...       ...       ...
437  0.041708  0.050680  0.019662  0.059744
438 -0.005515  0.050680 -0.015906 -0.067642
439  0.041708  0.050680 -0.015906  0.017282
440 -0.045472 -0.044642  0.039062  0.001215
441 -0.045472 -0.044642 -0.073030 -0.081414

[442 rows x 4 columns]

In [5]:
diabetes_X[["sex"]].drop_duplicates()

sex
0  0.050680
1 -0.044642

Here "sex" is categorical, so converting it to 0 & 1 

In [6]:
diabetes_X["sex"] = diabetes_X["sex"].replace(np.unique(diabetes_X["sex"])[0],0)
diabetes_X["sex"] = diabetes_X["sex"].replace(np.unique(diabetes_X["sex"])[1],1)

diabetes_X[["sex"]].drop_duplicates()

sex
0  1.0
1  0.0

In [7]:
X_train, X_test, y_train, y_test = train_test_split(diabetes_X, diabetes_y, test_size=0.3, random_state=42)

In [8]:
X_test = sm.add_constant(X_test)
X_train = sm.add_constant(X_train)

In [9]:
model = sm.OLS(y_train, X_train).fit()

In [10]:
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 target   R-squared:                       0.424
Model:                            OLS   Adj. R-squared:                  0.417
Method:                 Least Squares   F-statistic:                     55.98
Date:                Fri, 30 Sep 2022   Prob (F-statistic):           2.40e-35
Time:                        20:47:33   Log-Likelihood:                -1701.0
No. Observations:                 309   AIC:                             3412.
Df Residuals:                     304   BIC:                             3431.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        155.7696      4.734     32.904      0.0

**Question 5. b)**

The coefficients are:

In [11]:
x = pd.DataFrame(model.params).T
x

const        age        sex         bmi          bp
0  155.76959  49.848355 -10.227191  824.529801  416.295292

All the variables are normalized, except for 'sex' since it is already converted.

**age:** When the 'age' increases by 1,the predicted value increases by 49.85, all other variables are constant.


**sex:** 'sex' is categorical, when 'sex'=0 there are no changes. But, when 'sex'=1, the predicted value changes by 10.23.

**bmi:** For every value increase in 'bmi', the predicted value increases by 824.53, all other variables are constant.

**bp:** When the 'age' increases by 1 the predicted value increases by 416.29, all other variables are constant.

**Question 5. c)**

$Y_{pred} = \beta_{0} + \beta_{1}.X_{1} + \beta_{2}.X_{2} + \beta_{3}.X_{3} + \beta_{4}.X_{4}$

$Y_{pred} = 155.77 + (49.85*age) - (10.23*sex) + (824.53*bmi) + (416.29*bp)$

**Question 5. d)**

In [12]:
model_con = model.conf_int(alpha=0.05).T

In [13]:
for col in model_con.columns:
    print("The 95% confidence model of " + col + " is (" + str(round(model_con[col][0],3)) +"," + str(round(model_con[col][1],3)) + ")")

The 95% confidence model of const is (146.454,165.085)
The 95% confidence model of age is (-103.913,203.61)
The 95% confidence model of sex is (-23.928,3.473)
The 95% confidence model of bmi is (672.45,976.61)
The 95% confidence model of bp is (257.598,574.993)
